src/libsystemd/sd-daemon/sd-daemon.c

```diff
@@ -621,33 +633,36 @@ _public_ int sd_pid_notify_with_fds(
         int r;
 
         r = pid_notify_with_fds_internal(pid, state, fds, n_fds);
-
-        if (unset_environment)
-                assert_se(unsetenv("NOTIFY_SOCKET") == 0);
-
+        unsetenv_notify(unset_environment);
         return r;
 }
 
 _public_ int sd_pid_notify_barrier(pid_t pid, int unset_environment, uint64_t timeout) {
         _cleanup_close_pair_ int pipe_fd[2] = EBADF_PAIR;
         int r;
 
-        if (pipe2(pipe_fd, O_CLOEXEC) < 0)
-                return -errno;
+        r = RET_NERRNO(pipe2(pipe_fd, O_CLOEXEC));
+        if (r < 0)
+                goto finish;
 
-        r = sd_pid_notify_with_fds(pid, unset_environment, "BARRIER=1", &pipe_fd[1], 1);
+        r = pid_notify_with_fds_internal(pid, "BARRIER=1", &pipe_fd[1], 1);
         if (r <= 0)
-                return r;
+                goto finish;
 
         pipe_fd[1] = safe_close(pipe_fd[1]);
 
         r = fd_wait_for_event(pipe_fd[0], 0 /* POLLHUP is implicit */, timeout);
         if (r < 0)
-                return r;
-        if (r == 0)
-                return -ETIMEDOUT;
+                goto finish;
+        if (r == 0) {
+                r = -ETIMEDOUT;
+                goto finish;
+        }
 
-        return 1;
+        r = 1;
+finish:
```

Comment:
Instead, a cleanup helper in the form of utxent_cleanup can be introduced?

Reply:
dunno, it would feel weird no, because no variable to attach it to...

but maybe later, dunno. not generally opposed.